x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/Locate.java

```diff
@@ -136,12 +136,6 @@ public DataType dataType() {
 
     @Override
     public Expression replaceChildren(List<Expression> newChildren) {
-        if (start != null && newChildren.size() != 3) {
```

Comment:
Why this removal is _still_ ok given the new replaceChildrenSameSize method?

Reply:
Because the `replaceChildrenSameSize` guarantees that the number of the children won't change. If initially (using `new Locate(...)`) the node set the children correctly, it guarantees that the number of children won't change after. For example: If the `start` is null, it means that Locate has two children and `replaceChildrenSameSize` will check that `newChildren.size() == 2`. The `replaceChildren` is not called directly on `Locate`.

Reply:
Note: if you ever have to rewrite the children (and not just transform one-by-one when you should use `replaceChildrenSameSize`) of a Node, it is always preferred to use the constructor to create a new Node instead of the `replaceChildren` call. Unless you need an abstract constructor call (`replaceChildren`) to handle multiple subclasses without having to switch on the class (see `SimplifyCase`).

Reply:
>  it is always preferred to use the constructor to create a new Node instead of the replaceChildren call. 

What makes you say that?

Reply:
IMO if you already have a concrete class in your hand (specific rule for that particular class) with the children already destructured (meaning specific items in the children list accessible through named methods), it is more readable to call the constructor that expects the children as separate arguments than calling the `replaceChildren` with a list of expressions where you have to make sure anyways that the order in the list is in sync with how the `replaceChildren` destructures it and calls the same constructor that you would call. Having said that, I don't really see rules like this in our code. What I see is:

1. New instance is required becase some of the children in the subtree was changed by a transformation: `replaceChildrenSameSize` that calls the `replaceChildren` handles all the classes.
2. We want to transform the _varargs_ of multiple child classes the same way (for example `SimplifyConditional` rule): Calling the `replaceChildren` instead of switch on class + constructor call is better, `replaceChildren` already abstracts out the constructor call (order in the children list is already fixed by the parent class).
3. Rewriting whole expression trees (for example `CombineBinaryComparisons`): we create completely new expressions, change classes too, `replaceChildren` is not possible.